content/blog/2019-11-08-unified-theory-gc/index.md

```diff
@@ -0,0 +1,245 @@
+reference counts persist over time and are computed iteratively. The paper
+models reference counting collectors to keep a work list as tracing collectors
+do. Every time a pointer stored somewhere, its reference count is immediately
+incremented. Every time a pointer is overwritten or erased, it is added to the
+work list. Then, when the collection algorithm runs, the work list is iterated
+through, decrementing the reference count of each object in the list, and, if
+this causes the reference count to become zero, adding any objects that are
+pointed to by that object to the work list. When this algorithm terminates,
+every live object will still have a positive reference count, and the reference
+counts of dead objects could be zero (depending on if they are part of a garbage
+cycle).
+
+Real implementations of reference counting collection will typically not have a
+work list. Instead, they will immediately decrement the reference count whenever
+a pointer is overwritten, and, if the reference count becomes zero, recursively
+decrement the reference count of anything that is pointed to by that object. The
+algorithm is presented in this work-list format in order to more clearly
+represent its relationship to tracing garbage collection. Formulated in this
+way, if we look at the code for the central work-list algorithms for each type
+of garbage collector, we can see striking similarities between them:
+
+<img src="tracing-wl.png" style="width:100%"> | <img src="rc-wl.png" style="width:100%"> 
+:-------------------------------------:|:--------------------------:
+ Graph traversal algorithm for tracing | Work-list algorithm for RC
+
+The only differences between the two are that tracing increments the reference
+counts of objects on the worklist whereas RC decrements them, and that tracing
+checks if $\rho(w) = 1$ for when to add the objects that $w$ points to whereas
+RC checks if $\rho(w) = 0$. The two garbage collection strategies also differ in
+how $W$ is initialized. For tracing, $W$ initially contains all of the roots of
+the graph, $R$. For reference counting, $W$ starts with all of the objects that
+have had a pointer to them erased since the last time the algorithm ran. By
+viewing the two strategies through this formulation, we can see that they are
+opposites of each other in many ways:
+
+<img src="diff-v2.png" style="width:100%">
+
+As discussed in the mathematical model part, reference counting requires an extra pass to collect cycles. This is generally be done by two strategies: backup tracing, which traces (part of) the heap occasionally, and trial deletion, which attempts to decrease  $\rho(w)$ by trial-and-error guided by heuristics. However, notice that there is also counterpart for that in tracing: sweeping phase. In addition, while tracing converges to the fixed point value stating from the lower bound 0, reference counting starts from the upper bound, which is all incoming pointers that existed historically. 
```

Comment:
```suggestion
As discussed in the mathematical model part, reference counting requires an extra pass to collect cycles. This is generally be done by two strategies: backup tracing, which traces (part of) the heap occasionally, and trial deletion, which attempts to decrease  $\rho(w)$ by trial-and-error guided by heuristics. However, notice that there is also counterpart for that in tracing: the sweeping phase. In addition, while tracing converges to the fixed point value stating from the lower bound 0, reference counting starts from the upper bound, which is all incoming pointers that existed historically. 
```

Reply:
Perhaps instead of "the lower bound 0," consider "the lower bound $\varnothing$", if you mean the empty set?